### ISSUE TITLE
New F# features on .NET Framework

### ISSUE THREAD

**Issue by kasajian** [fsharp/fslang-design#640]
The C# language doesn't have very good features for .NET Framework 4.8 for new features. Some new features aren't even available unless you use .NET Core / .NET 5/6.

Is the F# language team making a similar decision, in that it's possible and likely for new F# language features to only work on the original .NET Framework (4.8 being the last version). But some new features will only target .NET 5/6. Had that happened already?

**Comment by dsyme** [comment]
New F# language features generally work if the underlying framework has the necessary functionality - we haven't deliberately restricted things.  

However there is no deliberate intention to keep designing features in a way that they light up on .NET Framework 4.8 and we certainly would expect some future features not to work on old frameworks.

**Comment by kasajian** [comment]
Thanks Don.   I mean, that certainly keeps your options open.    Why would you cripple the language simply because it cannot be supported (or supported efficiently) in a Framework that's no longer getting new features.

Having said that, because F# has been doing a lot of things for a long time without direct support from .NET Framework, it seems to me that new features are more likely to work with .NET Framework as compared to new Features in C#, which is more tightly tied to hip to the CLR.

In other words, in those cases when a green-field project has no choice but to target .NET Framework, the development team could stick with C# and expect to remain in the crippled world of "Most of C#8, and some of C#9, and probably that's it", or they could target F# and more likely remain current at least from the language features point of view.   Yes, it's a gamble.   I'm curious how right I will be.   Will check back in 5 years.

**Comment by smoothdeveloper** [comment]
@kasajian, I'm facing exactly those circumstances on features C# delivered, which obviously don't require new runtime than older .net framework (because you can hack your way around, but makes life miserable if you need to train people to deal with the hacks and no official support).

I think this hurts codebases, and effectively splits the language, not as strongly as python 2 -> 3 but still.

On F# side, it feels the adjustments to the language are life enhancing, but not trying to address shortcomings of earlier choices so much.

The soundness of older F# code doesn't pale severely when such code is brought into "most up to date green field code" as much as when you try to do this in C#, for those who'd like to use all the "non runtime dependent" enhancements, and you can't do it in existing .net framework codebases, and it feels mostly political decision (because even the technical cost would be IMO, warranted due to size of the existing codebases).

I can see this also causing problems for people trying to hire "C# developers" which will come in varying flavours due to those technical splits along the evolution of the language and the platform.

On the other end, I've not been severely impacted by F# not shipping support for new runtime features on day one, but observed it always caught up shortly after, where it mattered the most.

If we can keep it this way as to how it works with F#, that's great.

**Comment by kasajian** [comment]
@smoothdeveloper agreed.  I feel like if someone is stating a new project right now and must target .NET Framework, F# should be considered a better long-term option.  C# on .NET Framework seems all but abandoned,.

**Comment by dsyme** [comment]
I should point out, this is not particularly intentional. For example, we may well in the near-ish future decide to freeze the compiler on .NET Framework 4.x and develop/test only on .NET Core/.NET 5/.NET 6/.NET 7 and choose not to have new features work on .NET Framework.  I don't know - but I trust you see it would massively reduce our test matrix and simplify our life from an engineering perspective alone.

So please don't consider this as intentional.  It just happened to be the cheapest engineering path so far, but if at any point there had been significant costs or problems along the way, we would have chosen otherwise.

**Comment by cartermp** [comment]
The other thing is that we'll eventually have something in the language that's fundamentally incompatible with .NET Framework. Whether or not F# tooling goes out of its way to ensure that by default you don't get that stuff enabled will have to be decided. The cheapest option is to do nothing, and have .NET Framework users just observe a compile error and move on with their lives.

**Comment by dsyme** [comment]
Closing this discussion
